File: Backend/app/schemas/gate_exit.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

class GateExitMaterial(BaseModel):
    material_code: Optional[str] = None
    material_name: str
    quantity: float
    uom: str

class GateExitCreate(BaseModel):
    exit_type: str = Field(..., description="dispatch, jobwork_out, scrap, courier, return, transfer, sample")
    destination: str
    vehicle_no: Optional[str] = None
    driver_name: Optional[str] = None
    linked_document: Optional[str] = None
    customer: Optional[str] = None # Mapped to destination or remarks if needed, or we use destination for customer
    materials: List[GateExitMaterial] = []
    remarks: Optional[str] = None

class GateExitUpdate(BaseModel):
    exit_type: Optional[str] = None
    destination: Optional[str] = None
    vehicle_no: Optional[str] = None
    driver_name: Optional[str] = None
    linked_document: Optional[str] = None
    status: Optional[str] = None
    materials: Optional[List[GateExitMaterial]] = None
    remarks: Optional[str] = None

class GateExitResponse(BaseModel):
    id: str
    exit_number: str
    exit_type: str
    destination: str
    vehicle_no: Optional[str]
    driver_name: Optional[str]
    linked_document: Optional[str]
    status: str
    materials: List[GateExitMaterial]
    remarks: Optional[str]
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str]

    class Config:
        from_attributes = True

class GateExitStats(BaseModel):
    total_exits: int
    ready: int
    verified: int
    dispatched: int
    today_exits: int

File: Frontend/src/lib/api/gateExit.ts
import { apiClient } from './client';

export interface GateExitMaterial {
    material_code?: string;
    material_name: string;
    quantity: number;
    uom: string;
}

export interface GateExitCreate {
    exit_type: string;
    destination: string;
    vehicle_no?: string;
    driver_name?: string;
    linked_document?: string;
    materials: GateExitMaterial[];
    remarks?: string;
}

export interface GateExitResponse {
    id: string;
    exit_number: string;
    exit_type: string;
    destination: string;
    vehicle_no?: string;
    driver_name?: string;
    linked_document?: string;
    status: string;
    materials: GateExitMaterial[];
    remarks?: string;
    created_at: string;
}

export interface GateExitStats {
    total_exits: number;
    ready: number;
    verified: number;
    dispatched: number;
    today_exits: number;
}

export const gateExitApi = {
    create: async (data: GateExitCreate): Promise<GateExitResponse> => {
        return apiClient.post<GateExitResponse>('/gate-exits/', data);
    },

    list: async (params?: { limit?: number; offset?: number }): Promise<GateExitResponse[]> => {
        const queryParams = new URLSearchParams();
        if (params?.limit) queryParams.append('limit', params.limit.toString());
        if (params?.offset) queryParams.append('offset', params.offset.toString());

        return apiClient.get<GateExitResponse[]>(`/gate-exits/?${queryParams.toString()}`);
    },

    getStats: async (): Promise<GateExitStats> => {
        return apiClient.get<GateExitStats>('/gate-exits/stats');
    },

    get: async (id: string): Promise<GateExitResponse> => {
        return apiClient.get<GateExitResponse>(`/gate-exits/${id}`);
    }
};
